### ISSUE TITLE
New Containers not launching consistently when using --maxTestSessions

### ISSUE THREAD

**Issue by dgracehmh** [zalando/zalenium#322]
Hi, 
When running with 
```docker run --rm -ti --name zalenium -p 4444:4444 -p 5555:5555 \
     -v $(which docker):/usr/bin/docker \
     -v /var/run/docker.sock:/var/run/docker.sock \
     -v /tmp:/tmp/node/home/seluser/Downloads \
     -v /tmp:/home/seluser/Downloads \
     --privileged dosel/zalenium start --chromeContainers 4 --maxDockerSeleniumContainers 5 --videoRecordingEnabled false --firefoxContainers 0 --maxTestSessions 5 --debugEnabled true
```

I am seeing that new containers are not getting launch if there are still sessions available on an existing running container. 
In the logs it looks like this 
```12:05:01.061 DEBUG - Validating {chrome.switches=--no-sandbox --headless --disable-gpu, goog:chromeOptions={args=[test-type, --no-sandbox, --headless, --disable-gpu, test-type], extensions=[]}, record-screenshots=false, browserName=chrome, idle-timeout=180, version=} in node with capabilities {seleniumProtocol=WebDriver, acceptSslCerts=true, screen-resolution=1920x1080, tz=Europe/Berlin, browserName=chrome, maxInstances=1, screenResolution=1920x1080, resolution=1920x1080, version=62.0.3202.94, platform=LINUX}
12:05:01.061 DEBUG - [DS] Sessions still available.
12:05:01.062 DEBUG - [DS] There are sessions available for {chrome.switches=--no-sandbox --headless --disable-gpu, goog:chromeOptions={args=[test-type, --no-sandbox, --headless, --disable-gpu, test-type], extensions=[]}, record-screenshots=false, browserName=chrome, idle-timeout=180, version=}, wont start a new node yet.
``` 
<img width="1166" alt="screen shot 2017-11-15 at 11 05 32" src="https://user-images.githubusercontent.com/24386227/32833035-64b943b4-c9f5-11e7-88e3-ce6b686e600b.png">

Zalenium Image Version(s):  
3.7.1c
Docker Version:  
Docker version 17.09.0-ce, build afdb6d4
If using docker-compose, version:  
na
OS: 
OSX El Capitan
Docker Command to start Zalenium:  

```docker run --rm -ti --name zalenium -p 4444:4444 -p 5555:5555 \
     -v $(which docker):/usr/bin/docker \
     -v /var/run/docker.sock:/var/run/docker.sock \
     -v /tmp:/tmp/node/home/seluser/Downloads \
     -v /tmp:/home/seluser/Downloads \
     --privileged dosel/zalenium start --chromeContainers 4 --maxDockerSeleniumContainers 5 --videoRecordingEnabled false --firefoxContainers 0 --maxTestSessions 5 --debugEnabled true
```

## Expected Behavior -
With tests in the queue, Zalenium should not wait until all sessions on running containers are filled before launching new containers
## Actual Behavior -
If there are sessions available on 1 running container, new containers will not be launched even though there are tests in the queue

**Comment by straris** [comment]
Damn, I was wondering what was going on, that makes sense.
- Launching zalenium `docker run -d -ti --name zalenium -p 4444:4444     -v /var/run/docker.sock:/var/run/docker.sock     -v /tmp/videos:/home/seluser/videos     --privileged dosel/zalenium start --maxTestSessions 20 --chromeContainers 3 --firefoxContainers 0 --videoRecordingEnabled true --maxDockerSeleniumContainers 4 --keepOnlyFailedTests true`

- The tests will timeout as one only one node is created although there are tests pilling up in the queue.

**Comment by diemol** [comment]
Hi @dgracehmh, @straris,

> Zalenium should not wait until all sessions on running containers are filled before launching new containers

That is exactly the current behaviour written in the code right now :)  

The logic behind it is that if you want to run 20 tests, in 4 threads, using Chrome, then you start Zalenium they way you did (running max 5 tests per node):

```
dosel/zalenium start --chromeContainers 4 --maxDockerSeleniumContainers 5 \
    --videoRecordingEnabled false --firefoxContainers 0 --maxTestSessions 5
``` 

So it depends on the amount of tests you were running in that test run. 

Going back to the current logic, where it waits until all available sessions were used first, it was made with a conservative approach, to avoid creating tons of containers if they are not needed. So far it has been fine, but we are open to feedback and to the option of changing it to a more aggressive approach or something in between. This is just something that we need to think about.

**Comment by dgracehmh** [comment]
Hi @diemol . 
In the example above where you are running 20 tests in 4 threads using Chrome, that makes sense. Another scenario however is I want to run 200 tests, in 4 threads using Chrome. Some tests are faster than others etc. If there is one long running test running while all the sessions are finished on the other 3 containers, we have a queue of 3 tests waiting on the Grid for one test to finish on the only running container. That scenario to me seems like waste, the 3 waiting tests could easily be picked up and run on new containers. 
``` to avoid creating tons of containers if they are not needed``` if we are using ```--maxDockerSeleniumContainers 5``` we will have at most 5 containers running at any one time? Correct me if I am wrong.
Again, I really appreciate the time you guys have put in to this and also the time to answer my weird queries :)

**Comment by diemol** [comment]
Completely agreed in all your comments :) 
For that use case, the current approach is not 100% fitting since different tests take different time.

The was a suggestion back then, to have a new flag that mentions the amount of containers that should be always running, which fits your case.

I will leave this issue open, so we can throw some ideas and then look for a way to improve the current logic. Maybe you have an idea already?

**Comment by straris** [comment]
In my case using `--maxTestSessions 1` does the trick.
If understood correctly the bottomline  ( @diemol please correct me if not), the shorter your tests are, the more you can increase the number of sessions per container without a risk of timeout.

In my case most of the tests take more than a few seconds..

**Comment by diemol** [comment]
@straris 
`--maxTestSessions 1` is the default behaviour. Which creates a container for each test.
With the current logic, I guess what you say is true.

The main idea behind `--maxTestSessions` is to reuse a container. Having a brand new container for each test is quite nice, and it is ok with test suites with a few tests. But with larger test suites, creating a container per test could easily add between 3-10 seconds per test (depends on the machine), so these large test suites would have a shorter execution time.

**Comment by straris** [comment]
@diemol  exactly, having one of those large test suites and the execution time is greatly increased.
The flag for keeping some containers alive sounds like a good solution for this case.

**Comment by straris** [comment]
<img width="554" alt="screen shot 2017-11-15 at 17 46 00" src="https://user-images.githubusercontent.com/12101347/32848487-f403ddb8-ca2c-11e7-90ae-cc196b1cfc43.png">
Ouch.. this sometimes is enough time for a test to timeout.

**Comment by straris** [comment]
Correcting the above: The tests will always timeout waiting for a new container to be ready

**Comment by straris** [comment]
@diemol instead of a flag that keeps a number of nodes always running, would it make sense to have a timeout for running nodes? (wait for x time with no new test requests before stopping the nodes)

This should make sure nodes won't shut down if there are pending test requests.

**Comment by diemol** [comment]
@straris 
Containers coming up after 35-40 seconds usually means that you have many containers running for the hardware capabilities available. As mentioned in another issue, start with a couple of threads and go up slowly to fine the right amount for your setup.

About your most recent [comment](https://github.com/zalando/zalenium/issues/322#issuecomment-345221961), containers always shutdown after the max amount of sessions was reached. In the past we had something like an idle time for containers, and when it was reached the container was shutting down. It was not really adding a lot of value and it was kind of confusing for users.

**Comment by straris** [comment]
From the perspective of a large test suite running every hour I'm failing to see the use of test session limit. Specifically every time a new test is added, that means the dock container will have to be re-run and re-adjusted to the new number.

But perhaps that's not the most common use case.

**Comment by stefankahn** [comment]
Im suffering the same issue currently. Tests are in queue but no new containers are spawning until the current ones have finished with their test sessions. 

On a side note setting the number of chrome containers is only enforced on launch. It does not keep a minimum container count up.

**Comment by diemol** [comment]
@straris the idea is to save time, and I guess this is mostly needed for large test suites.
@stefankahn that is a good point, the idea was to start X number of containers when launching Zalenium and setting the `--maxTestSessions` to have them always running.

Perhaps removing the validation of "Sessions still available." will simplify things for most users, it could leave a few containers running at the end of the test suite, but I guess it should be ok. 

What do you think?

**Comment by stefankahn** [comment]
@diemol My 2 cents. I think your idea of test sessions per container is good as it gives a finite lifespan for the container. New containers should spawn up to your --maxDockerSeleniumContainers when a test will end up in queue, irrelevant of how many test sessions are still available.

This seems to be a bit of a change in view of how zalenium works. It shifts away from bringing zalenium up for your specific test run and shifts towards having it running as dedicated test infrastructure for anyone to hit anytime with the bonus of it being self spawning as it needs.

**Comment by diemol** [comment]
That same pseudocode is what I had in mind @stefankahn 

I just need to remove an `if` and should be good to go. I'll test it tomorrow and if it goes well, there'll be a release in a few days.

**Comment by diemol** [comment]
Hi @dgracehmh, @straris, @stefankahn 

This was just released https://github.com/zalando/zalenium/releases/tag/3.7.1e, please give it a try and give us some feedback.

The validation works almost like the pseudocode shown above. The slight change is when spawning a new container, we check first if a container is cleaning up (after running a test) then we wait a bit until it is ready. This is faster than spawning new containers always. And when all containers are being used, and a new test request comes, then a new one is spawned up.

I'll close this issue since it was released, please open new issues to if other problems show up.

**Comment by straris** [comment]
You guys rock

**Comment by dgracehmh** [comment]
Looks great so far @diemol , delighted with this!! Thank you.

**Comment by diemol** [comment]
Thanks for the feedback and nice words! 
To support us, don't forget to star the project if you have not done that yet :)

**Comment by dgracehmh** [comment]
Already starred.
btw, for initial runs it looks like my run time has come down from 1hr 57 minutes to 56 minutes, more than half. Thats with ```dosel/zalenium start --desiredContainers 3 --maxDockerSeleniumContainers 3 --maxTestSessions 4 --videoRecordingEnabled false```
Pretty damn cool!!

**Comment by diemol** [comment]
Wow, cool! How many tests are you running?

**Comment by dgracehmh** [comment]
There are about 170 tests in total, some pretty involved.

**Comment by stefankahn** [comment]
@diemol The change is good, I am no longer flooded with timeouts from tests sitting in queue during parallel runs :+1: 

Is a flag to keep the minimum node count >= --desiredContainers  flag possible as well?

My particular use case:
Running tests across 5 threads, Having 6 containers up removes all wait time.
We are using zalenium as an always up grid for any team to target, opposed to starting it for each test run.
